Solution to Assignment 4

fulopbbea opened the pull request:
Two solutions added:
-recursive flood fill algorithm
-(pseudo)linear equivalence classes based algorithm

The tests use large maps generated from binary images to exhaustively test each if branch in the implementation. (The number of islands isn't always straightforward due to 4-connectivity.)

fulopbbea committed:
Solution to Assignment 4

Two solutions added:
-recursive flood fill algorithm
-(pseudo)linear equivalence classes based algorithm

The tests use large maps generated from binary images to (hopefully) exhaustively test each if branch in the implementation. (The number of islands isn't always straightforward due to 4-connectivity.)

elifguler96 commented on src/assignment4/Assignment4.java:
```diff
@@ -0,0 +1,109 @@
+    }
+    
+    public static int countIslandsEquivalenceGraph(boolean[][] map, int height, int width) {
+        int[][] label = new int[height][width];
```
I don't really understand what this `label` array is for. Can you explain it with a comment?

elifguler96 commented on src/assignment4/Graph.java:
```diff
@@ -0,0 +1,74 @@
+        this.nodes.get(id2).addNeighbour(this.nodes.get(id1));
+    }
+    
+    private void depthFisrtSearch(GraphNode node) {
```
Small typo, should be `depthFirstSearch` 😄

elifguler96 reviewed (changes requested):
Looks good overall, added some small comments. Also one more thing: I think it is a great practice to add Javadoc comments to your methods, even though they are self-explanatory. I could suggest you add those. @brzozan can I get your opinion on that?

brzozan commented on test/assignment4/Assignment4Test.java:
```diff
@@ -0,0 +1,77 @@
+import org.junit.Test;
+import static org.junit.Assert.*;
+
+public class Assignment4Test {
```
Clever use of images!

More idiomatic way to handle these though would be via [parameterized tests](https://github.com/junit-team/junit4/wiki/parameterized-tests).

brzozan commented on test/assignment4/Assignment4Test.java:
```diff
@@ -0,0 +1,77 @@
+    }
+    
+    @BeforeClass
+    public static void setUpClass() {
```
No need to load images in `@BeforeClass` and store them in map. You could safely do it one by one from a helper method called by tests.

Always aim to have code as simple as possible, unless you have a very good reason to make it more complex :)

brzozan commented on src/assignment4/Assignment4.java:
```diff
@@ -0,0 +1,109 @@
+        for (int i = 0; i < height; i++) {
+            for (int j = 0; j < width; j++) {
+                if (map[i][j]) {
+                    if (i - 1 >= 0) {
```
These multi-level ifs are a bit difficult to follow. Could they be somehow split to separate methods?
Or at least have some helper variables for `i - 1` and `j - 1` explaining their meaning + some more comments to clarify what higher level ifs represent?

brzozan reviewed (changes requested):
Very nice PR with interesting approach to the problem and to the testing!

Just a few comments.

fulopbbea committed:
JavaDoc and comments added

-made the equivalence graph based solution a little bit more readable
-parameterized tests to follow

brzozan commented on src/assignment4/Assignment4.java:
```diff
@@ -2,6 +2,14 @@
+    /**
+     * Makes a copy of the map (boolean[][]) used by the functions that destroy the map while processing it.
+     * 
+     * @param map
```
If you include `@param` comments, please add descriptions of parameters. If you think they are self-explanatory, please remove `@param`s as they do not bring any value.

brzozan reviewed (approved):
LGTM, but please remove empty `@param` comments before merging.
